src/optimizer/transforms/combine-repeating-patterns-transform.js

```diff
@@ -0,0 +1,206 @@
+/**
+ * The MIT License (MIT)
+ * Copyright (c) 2017-present Dmitry Soshnikov <dmitry.soshnikov@gmail.com>
+ */
+
+'use strict';
+
+const NodePath = require('../../traverse/node-path');
+
+/**
+ * A regexp-tree plugin to combine repeating patterns.
+ *
+ * /^abcabcabc/ -> /^abc{3}/
+ * /^(?:abc){2}abc/ -> /^(?:abc){3}/
+ * /^abc(?:abc){2}/ -> /^(?:abc){3}/
+ */
+
+module.exports = {
+  Alternative(path) {
+    const {node} = path;
+
+    // We can skip the first child
+    let index = 1;
+    while (index < node.expressions.length) {
+      let child = path.getChild(index);
+      index = Math.max(1, combineRepeatingPatternLeft(path, child, index));
+
+      if (index >= node.expressions.length) {
+        break;
+      }
+
+      child = path.getChild(index);
+      index = Math.max(1, combineWithPreviousRepetition(path, child, index));
+
+      if (index >= node.expressions.length) {
+        break;
+      }
+
+      child = path.getChild(index);
+      index = Math.max(1, combineRepetitionWithPrevious(path, child, index));
+
+      index++;
+    }
+  }
+};
+
+// abcabc -> (?:abc){2}
+function combineRepeatingPatternLeft(alternative, child, index) {
+  const {node} = alternative;
+
+  const nbPossibleLengths = Math.ceil(index / 2);
+  let i = 0;
+
+  while (i < nbPossibleLengths) {
+    const startIndex = index - 2 * i - 1;
+    let right, left;
+
+    if (i === 0) {
+      right = child;
+      left = alternative.getChild(startIndex);
+    } else {
+      right = NodePath.getForNode({
+        type: 'Alternative',
+        expressions: [...node.expressions.slice(index - i, index), child.node]
+      });
+
+      left = NodePath.getForNode({
+        type: 'Alternative',
+        expressions: [...node.expressions.slice(startIndex, index - i)]
+      });
+    }
+
+    if (right.hasEqualSource(left)) {
+      for (let j = 0; j < 2 * i + 1; j++) {
+        alternative.getChild(startIndex).remove();
+      }
+
+      child.replace({
+        type: 'Repetition',
+        expression: i === 0 ? right.node : {
+          type: 'Group',
+          capturing: false,
+          expression: right.node
+        },
+        quantifier: {
+          type: 'Quantifier',
+          kind: 'Range',
+          from: 2,
+          to: 2,
+          greedy: true
+        }
+      });
+      return startIndex;
+    }
+
+    i++;
+  }
+
+  return index;
+}
+
+// (?:abc){2}abc -> (?:abc){3}
+function combineWithPreviousRepetition(alternative, child, index) {
+  const {node} = alternative;
+
+  let i = 0;
+  while (i < index) {
```

Comment:
Amount of loops is growing in the optimizer module :) Should be fine for now, but later we should be checking performance regressions. Usually it's OK on a smaller code base, but when you have thousand of files which you need to handle, smaller peaces can combine to some performance bottleneck.

Reply:
@DmitrySoshnikov: I agree. It's super hard to make smart optimizations while keeping the amount of loops low :D